lms/djangoapps/bookmarks/views.py

```diff
@@ -0,0 +1,289 @@
+"""
+HTTP end-points for the Bookmarks API.
+
+For more information, see:
+https://openedx.atlassian.net/wiki/display/TNL/Bookmarks+API
+"""
+import logging
+
+from django.core.exceptions import ObjectDoesNotExist
+from django.utils.translation import ugettext as _, ugettext_noop
+
+from rest_framework import status
+from rest_framework import permissions
+from rest_framework.authentication import OAuth2Authentication, SessionAuthentication
+from rest_framework.generics import ListCreateAPIView
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey, UsageKey
+
+from openedx.core.lib.api.permissions import IsUserInUrl
+from openedx.core.lib.api.serializers import PaginationSerializer
+
+from xmodule.modulestore.exceptions import ItemNotFoundError
+
+from lms.djangoapps.lms_xblock.runtime import unquote_slashes
+
+from . import DEFAULT_FIELDS, OPTIONAL_FIELDS, api
+from .serializers import BookmarkSerializer
+
+log = logging.getLogger(__name__)
+
+
+class BookmarksViewMixin(object):
+    """
+    Shared code for bookmarks views.
+    """
+
+    def fields_to_return(self, params):
+        """
+        Returns names of fields which should be included in the response.
+
+        Arguments:
+            params (dict): The request parameters.
+        """
+        optional_fields = params.get('fields', '').split(',')
+        return DEFAULT_FIELDS + [field for field in optional_fields if field in OPTIONAL_FIELDS]
+
+    def error_response(self, message, error_status=status.HTTP_400_BAD_REQUEST):
+        """
+        Create and return a Response.
+
+        Arguments:
+            message (string): The message to put in the developer_message
+                and user_message fields.
+            status: The status of the response. Default is HTTP_400_BAD_REQUEST.
+        """
+        return Response(
+            {
+                "developer_message": message,
+                "user_message": _(message)  # pylint: disable=translation-of-non-string
+            },
+            status=error_status
+        )
+
+
+class BookmarksView(ListCreateAPIView, BookmarksViewMixin):
+    """
+    **Use Case**
+
+        * Get a paginated list of bookmarks for a user.
+
+        The list can be filtered by passing parameter course_id to only
+        include bookmarks from a particular course.
+
+        The bookmarks are always sorted in descending order by creation date.
+
+        Each page in the list contains 10 bookmarks by default. The page
+        size can be altered by passing the parameter page_size.
+
+        To include the optional fields pass the values in 'fields' parameter. As
+        shown in example request.
+
+
+        * Post a new bookmark for particular block.
```

Comment:
This needs more detail about the patch format, and the possible error codes.

https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/user_api/accounts/views.py#L130